framework/decode/vulkan_offscreen_swapchain.cpp

```diff
@@ -0,0 +1,300 @@
+                                                         const QueueInfo*   queue_info,
+                                                         uint32_t           wait_semaphore_count,
+                                                         const VkSemaphore* wait_semaphores,
+                                                         uint32_t           signal_semaphore_count,
+                                                         const VkSemaphore* signal_semaphores,
+                                                         VkFence            fence)
+{
+    // Get the per swapchain resource data so we have access to the swapchain-specific information.
+    if (swapchain_resources_.find(swapchain_info->handle) == swapchain_resources_.end())
+    {
+        GFXRECON_LOG_ERROR("VulkanOffscreenSwapchain::SingalSemaphoresFence missing swapchain resource data for "
+                           "swapchain (ID = %" PRIu64 ")",
+                           swapchain_info->capture_id);
+        return VK_ERROR_UNKNOWN;
+    }
+
+    auto& swapchain_resources = swapchain_resources_[swapchain_info->handle];
+    assert(swapchain_resources != nullptr);
+
+    uint32_t queue_family_index = default_queue_family_index_;
+    if (queue_info)
+    {
+        queue_family_index = queue_info->family_index;
+    }
+    // Find the appropriate CommandCopyData struct for this queue family
+    if (swapchain_resources->copy_cmd_data.find(queue_family_index) == swapchain_resources->copy_cmd_data.end())
+    {
+        GFXRECON_LOG_ERROR(
+            "VulkanOffscreenSwapchain::SingalSemaphoresFence missing swapchain resource copy command data for "
+            "queue_family_index: %d in swapchain (ID = %" PRIu64 ")",
+            queue_family_index,
+            swapchain_info->capture_id);
+        return VK_ERROR_UNKNOWN;
+    }
+    auto& copy_cmd_data  = swapchain_resources->copy_cmd_data[queue_family_index];
+    auto  command_buffer = copy_cmd_data.command_buffers[capture_image_index];
+
+    VkCommandBufferBeginInfo begin_info = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO };
+    begin_info.pNext                    = nullptr;
+    begin_info.flags                    = 0;
+    begin_info.pInheritanceInfo         = nullptr;
+
+    VkResult result = device_table_->ResetCommandBuffer(command_buffer, 0);
+    if (result != VK_SUCCESS)
+    {
+        return result;
+    }
+    result = device_table_->BeginCommandBuffer(command_buffer, &begin_info);
+    if (result != VK_SUCCESS)
+    {
+        return result;
+    }
+    result = device_table_->EndCommandBuffer(command_buffer);
+    if (result != VK_SUCCESS)
+    {
+        return result;
+    }
+
+    VkPipelineStageFlags wait_stage  = VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT;
+    VkSubmitInfo         submit_info = { VK_STRUCTURE_TYPE_SUBMIT_INFO };
+    submit_info.commandBufferCount   = 1;
+    submit_info.pCommandBuffers      = &command_buffer;
```

Comment:
Why not just set this to nullptr and Count to zero here? Isn't this function a bit overkill when you could just do https://github.com/ARM-software/lavatube/blob/main/src/hardcode_read.cpp#L426 instead? Or am I missing something?

And why inject a vkQueueWaitIdle?

Reply:
Thank you for this information. I didn't know commandbuffer wasn't necessary. Removed. 

vkQueueWaitIdle was also removed. this commandbuffer was used some else where, so I needed to ensure the commandbuffer was done. But now, it isn't needed.